### PR TITLE
Rework optimizers

### PR DESCRIPTION
**Pull request by benikm91**
This PR builds on #137 and should be merged afterward.

I am updating the optimizers to run with deepwit GPT branch. By doing this I found an issue with the optimizers:

* The precision I added recently was suboptimal, as the optimizers took two generics [Params, V]; however, Params could be of a different precision than V, resulting in potential errors. I fixed this with the abstraction IsFloatTree that allows removing V from optimizers; additionally, 
* I removed beta1 and beta2 from AdamState, as these are implementation details and not the actual state of Adam. Storing the iteration is mathematically enough, as beta1^n allows for bias correction in exponential decay. However, beta1^n is an inefficient implementation. I implemented a stateful abstraction, SequenceFunction, to handle this. Please check if this is okay.
*  I added a primitive learning rate schedule from the lr-schedule branch of my fork, used in GPT-2 training. Check the design. Extension is possible later. Currently implemented only in Adam.

### CODE REVIEW

File: core/src/main/scala/dimwit/tensortree/FloatTree.scala [hunk 1]
```diff
@@ -138,3 +138,24 @@ object FloatTree:
         p.map([T <: Tuple] => (n: Labels[T]) ?=> (a: Tensor[T, V]) => a.asFloat(vtype)).asInstanceOf[F[NewV]]
 
 type FloatTreeFor[V] = [P] =>> FloatTree[P, V]
+
+/** A typeclass that proves P is a FloatTree, hiding the specific float type V
+  * from method signatures while keeping the evidence available.
+  */
+trait IsFloatTree[P]:
```
Comment (marcelluethi) [inline core/src/main/scala/dimwit/tensortree/FloatTree.scala:149]: I don't quite understand how `IsFloatTree` is different from `FloatTree`. Wouldn't it be possible to enforce the `IsFloating[V]` constraint there already?

File: core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala [hunk 2]
```diff
@@ -1,11 +1,54 @@
...
+  * }}}
+  * while allowing mathematically valid abstractions by exposing this stateful logic
+  * as a simple, pure `Long => Double` function to the consuming optimizer.
+  *
+  * ==Restriction==
+  * A SequenceFunction can only be used if we can guarantee calling with increasing step.
+  *
+  * @tparam R The type of the state.
+  */
+trait SequenceFunction[R] extends (Int => R):
```
Comment (marcelluethi) [inline core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala]: I am not an expert on optimizers and learning rate adaption in deep learning. But my naive guess would be that way more work is done within a computation step than for the learning rate adaption. If this is true, why would we need a "high performance" abstraction, rather than having a simple function?

File: core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala [hunk 3]
```diff
@@ -73,62 +112,74 @@ case class Lion(learningRate: Double, weightDecay: Double = 0.0f, beta1: Double
...
 )
 
 /** Implements the Adam optimization algorithm.
   *
   * @see [[https://arxiv.org/abs/1412.6980 Adam: A Method for Stochastic Optimization]]
   */
-case class Adam(
-    learningRate: Double, // step size (learning rate)
+class Adam(
+    learningRate: Double | (Int => Double),
```
Comment (marcelluethi) [inline core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala]: I don't understand this interface. I assume (Int => Double) allows for adapting the learning rate. But isn't this the same as the learning rate scheduler? How does serialization (for a checkpoint) work? I guess there should be more than a function for that. Should the learning rate somehow be stored with the state?

File: core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala [hunk 4]
```diff
@@ -73,62 +112,74 @@ case class Lion(learningRate: Double, weightDecay: Double = 0.0f, beta1: Double
...
   private val β1 = b1
   private val β2 = b2
+  private val ε = epsilon
+
+  private val betaBiasCorrection: Int => (Double, Double) = new SequenceFunction[(Double, Double)]:
+    private var β1ₜ = 1d; private var β2ₜ = 1d
+    def doStep(): Unit = β1ₜ *= β1; β2ₜ *= β2
+    def state(): (Double, Double) = (β1ₜ, β2ₜ)
+
+    override protected def outOfOrderErrorMessage(step: Long, current: Long): String =
```
Comment (marcelluethi) [inline core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala]: Is this a consequence of the stateful abstraction? If yes, is the price worth paying?

**Review by marcelluethi — COMMENTED** [review]
Thanks for starting the implementation of these ideal. I am not an expert in optimizer and I am not aware of all the practical issues. Therefore my review is mostly a set of questions I was wondering about when reading the code.

File: core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala (hunk 2 above)
Comment (benikm91) [reply core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala]: Yes, it will not matter computation-wise. It felt more like bad code if exponential decay re-evaluates each iteration with .pow; however, mathematically, this is what happens. I changed to .pow in the last commit, until SequenceFunction will show to be necessary.

File: core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala (hunk 3 above)
Comment (benikm91) [reply core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala]: I changed (Int => Double) to LearningRateSchedule type alias to make it more explicit.

File: core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala (hunk 4 above)
Comment (benikm91) [reply core/src/main/scala/dimwit/optimizer/GradientOptimizer.scala]: Yes, it is a consequence. No, probably not worth paying until shown to be necessary due to slowdown.

File: core/src/main/scala/dimwit/tensortree/FloatTree.scala (hunk 1 above)
Comment (benikm91) [reply core/src/main/scala/dimwit/tensortree/FloatTree.scala:149]: FloatTree[P, V] is for a specific V.
IsFloatTree[P] marks any possible FloatTree

I can write

```scala
def x[P[_], V](p: P[V])(using FloatTree[P, V])
```

But then we can't use for hard-coded precision in e.g. params:

```scala
// x not applicable to as Params1 is not a P[_] type (takes no generic)
case class Params1(weights: Tensor2[Row, Col, Float32]

// x only applicable to as Params2 is a P[_] type
case class Params2[V: IsFloating](weights: Tensor2[Row, Col, V]
```

If I do 

```scala
def x2[P](p: P)(using IsFloatTree)

// x2 applicable to Params1[Float32] and applicable to Params2.
```

So far to the motivation. I don't know if there is a better solution :) Best I came up with.

Comment (benikm91) [reply core/src/main/scala/dimwit/tensortree/FloatTree.scala:149]: To make it concrete for the optimizers. `IsFloatTree` was here necessary to make the VAE example run that has hard coded Params precision. I think we should support hard coding precision.

Comment (marcelluethi) [reply core/src/main/scala/dimwit/tensortree/FloatTree.scala:149]: Do we need the higher kinded type here? Maybe something like would be easier to work with? 
``` 
def x[P, V](p: P)(using FloatTree[P, V])
```

**Review by marcelluethi — APPROVED** [review]
Thanks for the careful consideration of the issues. I think the solution you found at the end makes total sense and resolves both the issues we started with. 

Please squash the commits before/while merging.
